File: frontend/src/types/mikrotik.ts
export interface MikrotikDevice {
  id: number;
  name: string;
  ip_address: string;
  port: number;
  username: string;
  password: string;
  group_id?: number;
  group_name: string;
  location?: string;
  description?: string;
  router_board?: string;
  architecture?: string;
  version?: string;
  build_time?: string;
  is_online: boolean;
  last_seen?: string;
  connection_attempts: number;
  last_error?: string;
  cpu_load?: string;
  uptime?: string;
  created_at: string;
  updated_at: string;
}

export interface MikrotikDeviceCreate {
  name: string;
  ip_address: string;
  port: number;
  username: string;
  password: string;
  group_id?: number;
  group_name: string;
  location?: string;
  description?: string;
  router_board?: string;
}

export interface MikrotikDeviceUpdate {
  name?: string;
  ip_address?: string;
  port?: number;
  username?: string;
  password?: string;
  group_id?: number;
  group_name?: string;
  location?: string;
  description?: string;
  router_board?: string;
  manual_override?: boolean;
}

export interface DeviceInterface {
  id: number;
  device_id: number;
  name: string;
  type?: string;
  mac_address?: string;
  running: boolean;
  disabled: boolean;
  rx_bytes: number;
  tx_bytes: number;
  rx_packets: number;
  tx_packets: number;
  last_updated: string;
}

export interface DeviceLog {
  id: number;
  device_id: number;
  log_level: string;
  message: string;
  command?: string;
  response?: any;
  timestamp: string;
}

export interface DeviceCommand {
  id: number;
  device_id: number;
  command: string;
  status: string;
  result?: any;
  error_message?: string;
  created_at: string;
  started_at?: string;
  completed_at?: string;
  user_id?: string;
  user_name?: string;
}

export interface DeviceCommandCreate {
  device_id: number;
  command: string;
  user_id?: string;
  user_name?: string;
}

export interface ConnectionTestResult {
  success: boolean;
  message: string;
  device_info?: any;
}

export interface BulkOperationResult {
  success_count: number;
  failed_count: number;
  total_count: number;
  results: Array<{
    device_id: number;
    device_name: string;
    success: boolean;
    error?: string;
  }>;
}

export interface HighCpuDevice {
  id: number;
  name: string;
  ip_address: string;
  cpu_load: string;
  group_name: string;
  is_online: boolean;
  last_seen?: string;
}

export interface DeviceStats {
  total_devices: number;
  online_devices: number;
  offline_devices: number;
  groups: { [key: string]: number };
  recent_logs: DeviceLog[];
  high_cpu_devices: HighCpuDevice[];
}

export interface WebSocketMessage {
  type: string;
  device_id?: number;
  name?: string;
  is_online?: boolean;
  last_seen?: string;
  last_error?: string;
  message?: string;
}

export interface DeviceFilters {
  search?: string;
  group_name?: string;
  is_online?: boolean;
  skip?: number;
  limit?: number;
} 

File: backend/app/schemas/mikrotik.py
from pydantic import BaseModel, Field, validator
from typing import Optional, List, Dict, Any, TYPE_CHECKING
from datetime import datetime
import ipaddress

if TYPE_CHECKING:
    from . import Group, Subnet

class CredentialBase(BaseModel):
    name: str = Field(..., description="Display name for credential")
    username: str = Field(..., description="Username")
    password: str = Field(..., description="Password")
    description: Optional[str] = Field(None, description="Description")
    is_default: bool = Field(False, description="Is default credential")

class CredentialCreate(CredentialBase):
    pass

class CredentialUpdate(BaseModel):
    name: Optional[str] = None
    username: Optional[str] = None
    password: Optional[str] = None
    description: Optional[str] = None
    is_default: Optional[bool] = None

class Credential(CredentialBase):
    id: int
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True

class SubnetBase(BaseModel):
    name: str = Field(..., description="Display name for subnet")
    network: str = Field(..., description="Network IP address")
    cidr: int = Field(..., ge=1, le=32, description="CIDR notation")
    is_active: bool = Field(True, description="Is subnet active")

class SubnetCreate(SubnetBase):
    pass

class SubnetUpdate(BaseModel):
    name: Optional[str] = None
    network: Optional[str] = None
    cidr: Optional[int] = Field(None, ge=1, le=32)
    is_active: Optional[bool] = None

class Subnet(SubnetBase):
    id: int
    created_at: datetime
    updated_at: datetime
    groups: Optional[List[Dict[str, Any]]] = None
    
    @property
    def network_info(self) -> Dict[str, Any]:
        """Calculate network information"""
        try:
            network = ipaddress.IPv4Network(f"{self.network}/{self.cidr}", strict=False)
            return {
                "first_ip": str(network.network_address),
                "last_ip": str(network.broadcast_address),
                "total_ips": int(network.num_addresses),
                "usable_ips": int(network.num_addresses) - 2 if network.num_addresses > 2 else 0,
                "network_address": str(network.network_address),
                "broadcast_address": str(network.broadcast_address),
                "netmask": str(network.netmask)
            }
        except:
            return {
                "first_ip": "Invalid",
                "last_ip": "Invalid", 
                "total_ips": 0,
                "usable_ips": 0,
                "network_address": "Invalid",
                "broadcast_address": "Invalid",
                "netmask": "Invalid"
            }
    
    class Config:
        from_attributes = True

class GroupBase(BaseModel):
    name: str = Field(..., description="Display name for group")
    description: Optional[str] = Field(None, description="Group description")
    color: str = Field("#1976d2", description="Hex color code for UI")
    is_active: bool = Field(True, description="Is group active")

class GroupCreate(GroupBase):
    pass

class GroupUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    color: Optional[str] = None
    is_active: Optional[bool] = None

class Group(GroupBase):
    id: int
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True

class MikrotikDeviceBase(BaseModel):
    name: str = Field(..., description="Device name")
    ip_address: str = Field(..., description="IP address")
    port: int = Field(8728, description="RouterOS API port")
    username: str = Field(..., description="Username")
    password: str = Field(..., description="Password")
    group_id: Optional[int] = Field(None, description="Group ID")
    group_name: str = Field("default", description="Group name")
    location: Optional[str] = Field(None, description="Location")
    description: Optional[str] = Field(None, description="Description")
    router_board: Optional[str] = Field(None, description="Router board model")
    
    @validator('ip_address')
    def validate_ip(cls, v):
        import ipaddress
        try:
            ipaddress.ip_address(v)
        except ValueError:
            raise ValueError('Invalid IP address format')
        return v
    
    @validator('port')
    def validate_port(cls, v):
        if not (1 <= v <= 65535):
            raise ValueError('Port must be between 1 and 65535')
        return v

class MikrotikDeviceCreate(MikrotikDeviceBase):
    pass

class MikrotikDeviceUpdate(BaseModel):
    name: Optional[str] = None
    ip_address: Optional[str] = None
    port: Optional[int] = None
    username: Optional[str] = None
    password: Optional[str] = None
    group_id: Optional[int] = None
    group_name: Optional[str] = None
    location: Optional[str] = None
    description: Optional[str] = None
    router_board: Optional[str] = None
    manual_override: Optional[bool] = None

class MikrotikDevice(MikrotikDeviceBase):
    id: int
    router_board: Optional[str] = None
    architecture: Optional[str] = None
    version: Optional[str] = None
    build_time: Optional[str] = None
    manual_override: Optional[bool] = False
    is_online: bool = False
    last_seen: Optional[datetime] = None
    connection_attempts: int = 0
    last_error: Optional[str] = None
    cpu_load: Optional[str] = None
    uptime: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True

class DeviceInterface(BaseModel):
    id: int
    device_id: int
    name: str
    type: Optional[str] = None
    mac_address: Optional[str] = None
    running: bool = False
    disabled: bool = False
    rx_bytes: int = 0
    tx_bytes: int = 0
    rx_packets: int = 0
    tx_packets: int = 0
    last_updated: datetime
    
    class Config:
        from_attributes = True

class DeviceLog(BaseModel):
    id: int
    device_id: int
    log_level: str
    message: str
    command: Optional[str] = None
    response: Optional[Dict[str, Any]] = None
    timestamp: datetime
    
    class Config:
        from_attributes = True

class DeviceCommand(BaseModel):
    id: int
    device_id: int
    command: str
    status: str
    result: Optional[Dict[str, Any]] = None
    error_message: Optional[str] = None
    created_at: datetime
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    user_id: Optional[str] = None
    user_name: Optional[str] = None
    
    class Config:
        from_attributes = True

class DeviceCommandCreate(BaseModel):
    device_id: int
    command: str
    user_id: Optional[str] = None
    user_name: Optional[str] = None

class ConnectionTestResult(BaseModel):
    success: bool
    message: str
    device_info: Optional[Dict[str, Any]] = None

class BulkOperationResult(BaseModel):
    success_count: int
    failed_count: int
    total_count: int
    results: List[Dict[str, Any]]

class HighCpuDevice(BaseModel):
    id: int
    name: str
    ip_address: str
    cpu_load: str
    group_name: str
    is_online: bool
    last_seen: Optional[datetime] = None

    class Config:
        from_attributes = True

class DeviceStats(BaseModel):
    total_devices: int
    online_devices: int
    offline_devices: int
    groups: Dict[str, int]
    recent_logs: List[DeviceLog]
    high_cpu_devices: List[HighCpuDevice]

 